main/tests/test_class_db.cpp

```diff
@@ -0,0 +1,875 @@
+
+			exposed_class.constants.push_back(constant);
+		}
+
+		r_context.exposed_classes.insert(class_name, exposed_class);
+		class_list.pop_front();
+	}
+
+	TEST_END();
+}
+
+void add_builtin_types(Context &r_context) {
+	// NOTE: We don't care about the size and sign of int and float in these tests
+	for (int i = 0; i < Variant::VARIANT_MAX; i++) {
+		r_context.builtin_types.push_back(Variant::get_type_name(Variant::Type(i)));
+	}
+
+	r_context.builtin_types.push_back(_STR(Variant));
+	r_context.builtin_types.push_back(r_context.names_cache.vararg_stub_type);
+	r_context.builtin_types.push_back("void");
+}
+
+void add_global_enums(Context &r_context) {
+	int global_constants_count = GlobalConstants::get_global_constant_count();
+
+	if (global_constants_count > 0) {
+		for (int i = 0; i < global_constants_count; i++) {
+			StringName enum_name = GlobalConstants::get_global_constant_enum(i);
+
+			if (enum_name != StringName()) {
+				ConstantData constant;
+				constant.name = GlobalConstants::get_global_constant_name(i);
+				constant.value = GlobalConstants::get_global_constant_value(i);
+
+				EnumData enum_;
+				enum_.name = enum_name;
+				List<EnumData>::Element *enum_match = r_context.global_enums.find(enum_);
+				if (enum_match) {
+					enum_match->get().constants.push_back(constant);
+				} else {
+					enum_.constants.push_back(constant);
+					r_context.global_enums.push_back(enum_);
+				}
+			}
+		}
+
+		for (List<EnumData>::Element *E = r_context.global_enums.front(); E; E = E->next()) {
+			r_context.enum_types.push_back(E->get().name);
+		}
+	}
+
+	// HARDCODED
+	List<StringName> hardcoded_enums;
+	hardcoded_enums.push_back("Vector2.Axis");
+	hardcoded_enums.push_back("Vector2i.Axis");
+	hardcoded_enums.push_back("Vector3.Axis");
+	hardcoded_enums.push_back("Vector3i.Axis");
+	for (List<StringName>::Element *E = hardcoded_enums.front(); E; E = E->next()) {
+		// These enums are not generated and must be written manually (e.g.: Vector3.Axis)
+		// Here, we assume core types do not begin with underscore
+		r_context.enum_types.push_back(E->get());
+	}
+}
+
+TestResult run_class_db_tests() {
+	TEST_START();
+
+	Context context;
+
+	TEST_CHECK_FATAL(add_exposed_classes(context));
+	add_builtin_types(context);
+	add_global_enums(context);
+
+	const ExposedClass *object_class = context.find_exposed_class(context.names_cache.object_class);
+	TEST_FAIL_COND_FATAL(!object_class, "Object class not found.");
+	TEST_FAIL_COND_FATAL(object_class->base != StringName(),
+			"Object class derives from another class: '" + object_class->base + "'.");
+
+	for (ExposedClasses::Element E = context.exposed_classes.front(); E; E = E.next()) {
+		TEST_CHECK(validate_class(context, E.value()));
+	}
+
+	TEST_END();
+}
+
+MainLoop *test() {
+	TestResult pass = run_class_db_tests();
+
+	OS::get_singleton()->print("ClassDB tests: %s\n", pass == TestResult::PASS ? "PASS" : "FAILED");
+
+	if (pass == TestResult::FAILED) {
+		exit(1);
+	}
```

Comment:
I'm not sure that's correct, this will force closing the process without properly calling the destructors, no?

How about `OS::get_singleton()->set_exit_code(1);` instead and letting `main.cpp` finish gracefully?